language/diem-vm/parallel-executor/src/proptest_types/types.rs

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) The Diem Core Contributors
```

Comment:
Having these proptests is great! I have a couple questions about testing (both proptests, and more generally):
1. How are dependencies created in the proptest (/are they?)
2. It doesn't appear as though interleaved reads and writes and how the scheduler handles dependencies and dependency resolution is modelled in proptest. Is this an issue? This then leads to my last question; 
3. I currently don't see any non-proptest tests, and I'm wondering if having some hand written tests to explore esoteric cases and edge cases in the scheduler would be a good idea.

Reply:
> How are dependencies created in the proptest (/are they?)

They are created totally at random. Each transaction picks 10 keys they can write to from a key universe, and we generate the values that we want to write as well. With a list of transactions there will natuarally be dependencies in between those transactions as the key universe size is smaller comparing to the total # of transactions.

> It doesn't appear as though interleaved reads and writes and how the scheduler handles dependencies and dependency resolution is modelled in proptest. Is this an issue?

That goes to my previous point. Currently the setup is as following: we create 3000 transactions to be executed in parallel, each transaction will modify 10 keys and there will be a totally of 100 keys. 

> I currently don't see any non-proptest tests, and I'm wondering if having some hand written tests to explore esoteric cases and edge cases in the scheduler would be a good idea.

That would be a great idea. I will write test cases for those scenarios as a starter:
1. cycle keys: we will have X transactions writing to one key one after another, and Y distinct keys.
2. all-to-one dependency: X transactions writing to distinct keys and have one transaction reading from all of them.
3. tests around abort/skiprest mechanisms.